src/core/core.scale.js

```diff
@@ -314,39 +403,35 @@ module.exports = Element.extend({
 	},
 	calculateTickRotation: function() {
 		var me = this;
-		var context = me.ctx;
-		var tickOpts = me.options.ticks;
-		var labels = labelsFromTicks(me._ticks);
-
-		// Get the width of each grid by calculating the difference
-		// between x offsets between 0 and 1.
-		var tickFont = helpers.options._parseFont(tickOpts);
-		context.font = tickFont.string;
+		var options = me.options;
+		var tickOpts = options.ticks;
+		var ticks = me.getTicks();
+		var labelSizes, maxLabelWidth, maxLabelHeight, tickWidth, maxHeight, maxLabelDiagonal;
 
 		var labelRotation = tickOpts.minRotation || 0;
 
-		if (labels.length && me.options.display && me.isHorizontal()) {
-			var originalLabelWidth = helpers.longestText(context, tickFont.string, labels, me.longestTextCache);
-			var labelWidth = originalLabelWidth;
-			var cosRotation, sinRotation;
-
-			// Allow 3 pixels x2 padding either side for label readability
-			var tickWidth = me.getPixelForTick(1) - me.getPixelForTick(0) - 6;
-
-			// Max label rotation can be set or default to 90 - also act as a loop counter
-			while (labelWidth > tickWidth && labelRotation < tickOpts.maxRotation) {
-				var angleRadians = helpers.toRadians(labelRotation);
-				cosRotation = Math.cos(angleRadians);
-				sinRotation = Math.sin(angleRadians);
-
-				if (sinRotation * originalLabelWidth > me.maxHeight) {
-					// go back one step
-					labelRotation--;
-					break;
+		if (me._isVisible() && tickOpts.display) {
+			labelSizes = me._labelSizes = computeLabelSizes(me.ctx, parseTickFontOptions(tickOpts), ticks, me.longestTextCache);
+
+			if (ticks.length > 1 && me.isHorizontal()) {
+				maxLabelWidth = labelSizes.widest.width;
+				maxLabelHeight = labelSizes.highest.height - labelSizes.highest.offset;
+
+				// Estimate the width of each grid based on the canvas width, the maximum
+				// label width and the number of tick intervals
+				tickWidth = (me.chart.width - maxLabelWidth) / (ticks.length - 1);
+
+				// Allow 3 pixels x2 padding either side for label readability
+				if (maxLabelWidth + 6 > tickWidth) {
+					maxHeight = me.maxHeight - getTickMarkLength(options.gridLines)
+						- tickOpts.padding - getScaleLabelHeight(options.scaleLabel);
+					maxLabelDiagonal = Math.sqrt(maxLabelWidth * maxLabelWidth + maxLabelHeight * maxLabelHeight);
+					labelRotation = helpers.toDegrees(Math.min(
+						Math.asin(Math.min((labelSizes.highest.height + 6) / tickWidth, 1)),
+						Math.asin(Math.min(maxHeight / maxLabelDiagonal, 1)) - Math.asin(maxLabelHeight / maxLabelDiagonal)
+					));
+					labelRotation = Math.max(Number.MIN_VALUE, tickOpts.minRotation, Math.min(tickOpts.maxRotation, labelRotation));
```

Comment:
`calculateTickRotation` is called 2 times per `update` in the layout system, and if `labelRotation` is set to `0` in the first call, the left and right padding are not correctly added even if the second `calculateTickRotation` call sets it to the right value. So, this might be tricky, but some tiny number is added if the rotation is necessary.

Reply:
I don't think this `Number.MIN_VALUE` is strictly needed anymore (#6021). Would it cause extra padding in some situation? Also can you think of a regression test for this situation?

Reply:
Removed the check for `Number.MIN_VALUE`. I agree that it is no longer needed as `maxHeight` passed to `scale.update` doesn't change between the first and second calls, and the calculated rotation and padding values don't change either. Added a simple regression test.